test/jdk/java/awt/Scrollbar/AquaLFScrollbarTest/ScrollBarBorderTest.java

```diff
@@ -0,0 +1,221 @@
+import java.awt.Button;
+import java.awt.Scrollbar;
+import java.awt.Component;
+import java.awt.Graphics;
+import java.awt.Color;
+import java.awt.Insets;
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+
+/*
+ * @test
+ * @bug 8190264
+ * @summary JScrollBar ignores its border when using macOS Mac OS X Aqua look and feel
+ * @run main/manual ScrollBarBorderTest
+ */
+public class ScrollBarBorderTest implements ActionListener {
+
+    // On macOS 10.12.6 using the Mac look and feel (com.apple.laf.AquaLookAndFeel)
+    // the scroll bar ignores the custom border and allows the scroll thumb to move
+    // beneath the border. Run with:
+    // java ScrollBarBorderTest
+
+    // If run using any other look and feel (e.g. Metal) then the right side of
+    // the scroll bar stops at the border as expected. Run with:
+    // java -Dswing.defaultlaf=javax.swing.plaf.metal.MetalLookAndFeel ScrollBarBorderTest
+
+    // Java version: 1.8.0_151
+
+    private static JScrollBar scrollBar;
+    private static JPanel panel;
+    private static JFrame frame;
+    private static Frame instructionFrame;
+    private static GridBagLayout layout;
+    private static Panel mainControlPanel;
+    private static Panel resultButtonPanel;
+    private static TextArea instructionTextArea;
+    private static Button passButton;
+    private static Button failButton;
+    private static Thread mainThread = null;
+    private static boolean testPassed = false;
+    private static boolean isInterrupted = false;
+    private static final int testTimeOut = 300000;
+    private static String testFailMessage = "Test Failed. Thumb was able to move into border.";
+
+
+    public void createAndShowGUI() {
+        // create scroll bar
+        scrollBar = new JScrollBar(Scrollbar.HORIZONTAL);
+        scrollBar.setBorder(new CustomBorder());
+
+        // create panel
+        panel = new JPanel();
+        panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));
+        panel.setBorder(new EmptyBorder(20, 20, 20, 20));
+        panel.add(new JLabel(UIManager.getLookAndFeel().toString()));
+        panel.add(Box.createVerticalStrut(20));
+        panel.add(scrollBar);
+
+        // create frame
+        frame = new JFrame("ScrollBarBorderTest");
+        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+        frame.getContentPane().add(panel);
+        frame.pack();
+        frame.setVisible(true);
+    }
+
+    private void createInstructionUI() {
+        instructionFrame = new Frame("Scrollbar");
+        layout = new GridBagLayout();
+        mainControlPanel = new Panel(layout);
+        resultButtonPanel = new Panel(layout);
+
+        GridBagConstraints gbc = new GridBagConstraints();
+        String instructions
+                = "\nINSTRUCTIONS:\n"
+                + "\n   Try to drag the thumb of the scrollbar into the red zone."
+                + "\n   If the thumb is able to go into the red zone, click fail."
```

Comment:
If you place the thumb to the red zone by the code, will it be possible to check that by pixel color?

Reply:
Just updated the automation. I added a robot to try to move the thumb into the red zone, then had the robot click again to see if the thumb is under the cursor. If the behavior is correct, the thumb should have stopped before entering the red zone and scrollbar mouseListener should only detect 1 click. Otherwise it will detect 2 clicks (thumb is under cursor) and test fails. Is my logic correct?

Reply:
But it is not necessary to use the robot at all? Can you try to set the scollbar thumb position/move it programmatically w/o a robot? If it is possible then it will be rendered on top of the red area -> it will be possible to render it to the buffered image and just check the pixel in the right position(is it red or not).

Reply:
Is using robot to get the pixel color alright?

Reply:
You can render directly to the BufferedImage instead of the JFrame, so in this case, it will not be necessary to use a robot, you will be able to check the pixel directly.

Reply:
For example, you can take a look at the JButtonPaintNPE or TestBadBreak tests.

Reply:
In the TestBadBreak example, how do you know it's properly painting onto the final BufferedImage? When I try the same it doesn't seem to be painting anything on the BufferedImage, so there's no colors to compare.

Reply:
> In the TestBadBreak example, how do you know it's properly painting onto the final BufferedImage? When I try the same it doesn't seem to be painting anything on the BufferedImage, so there's no colors to compare.

How do you paint the frame into BufferedImage?
You may need to set the size of the frame explicitly.

There are many tests which paint components to the image. You can also take a look at test/jdk/javax/swing/text/FlowView/6318524/bug6318524.java and test/jdk/javax/swing/text/ParagraphView/6364882/bug6364882.java. Each of these two provides an option to save the BufferedImage to a file.

Reply:
> How do you paint the frame into BufferedImage?
You may need to set the size of the frame explicitly.

I'm overriding the paintComponent method of JPanel to instead directly paint onto a BufferedImage that I passed in, similar to the TestBadBreak example.

I pushed what I have so far, but it seems like the BufferedImage is blank

Reply:
The image is blank because this method is never called. `TestBadBreak` relies on the frame being shown, it adds the component into the frame and shows the frame to capture what's painted. You don't add the `panel` into the frame, therefore it's never painted.

Try to paint into the image directly, after you configured the panel, that is after `panel.add(scrollBar);`, call:

```java
Graphics2D graphics2D = image.createGraphics();
panel.paint(graphics2D);
graphics2D.dispose();
```

Then you can analyse the colour of pixels on the image just like you do now with robot.